segmentstore/server/src/test/java/io/pravega/segmentstore/server/containers/DebugStreamSegmentContainerTests.java

```diff
@@ -381,45 +384,53 @@ public void testDataRecoveryContainerLevel() throws Exception {
-        }
+        ContainerRecoveryUtils.createBackUpMetadataSegments(storage, containerCount, executorService(), TIMEOUT)
+                .thenCompose(backUpMetadataSegments -> {
+                    OperationLogFactory localDurableLogFactory2 = new DurableLogFactory(NO_TRUNCATIONS_DURABLE_LOG_CONFIG,
+                            new InMemoryDurableDataLogFactory(MAX_DATA_LOG_APPEND_SIZE, executorService()), executorService());
+                    // Starts a DebugSegmentContainer with new Durable Log
+                    @Cleanup
+                    TestContext context1 = createContext(executorService());
+
+                    @Cleanup
+                    MetadataCleanupContainer container2 = new MetadataCleanupContainer(containerId, CONTAINER_CONFIG, localDurableLogFactory2,
+                            context1.readIndexFactory, context1.attributeIndexFactory, context1.writerFactory, storageFactory,
+                            context1.getDefaultExtensions(), executorService());
+                    container2.startAsync().awaitRunning();
+                    Map<Integer, DebugStreamSegmentContainer> debugStreamSegmentContainersMap = new HashMap<>();
+                    debugStreamSegmentContainersMap.put(containerId, container2);
+
+                    try {
+                        // 4. Recover all segments.
+                        recoverAllSegments(storage, debugStreamSegmentContainersMap, executorService(), TIMEOUT);
+                        // 5. Update core attributes using back up segments.
+                        updateCoreAttributes(backUpMetadataSegments, debugStreamSegmentContainersMap, executorService(), TIMEOUT);
+                    } catch (Exception e) {
+                        log.error("Error while executing the test.", e);
+                        Assert.fail("Error while executing the test.");
+                    }
+
+                    // 6. Verify Segment Data.
+                    for (val sc : segmentContents.entrySet()) {
+                        // Contents.
+                        byte[] expectedData = sc.getValue().array();
+                        byte[] actualData = new byte[expectedData.length];
+                        container2.read(sc.getKey(), 0, actualData.length, TIMEOUT).join().readRemaining(actualData, TIMEOUT);
+                        Assert.assertArrayEquals("Unexpected contents for " + sc.getKey(), expectedData, actualData);
+
+                        // Length.
+                        val si = container2.getStreamSegmentInfo(sc.getKey(), TIMEOUT).join();
+                        Assert.assertEquals("Unexpected length for " + sc.getKey(), expectedData.length, si.getLength());
+
+                        // Attributes.
+                        val attributes = container2.getAttributes(sc.getKey(), Collections.singleton(attributeReplace), false, TIMEOUT).join();
```

Comment:
Joining a future inside a future does not look great. Actually, is really needed this change in the test? Could you just `get()` the future related to `containerRecoveryUtils.createBackUpMetadataSegments()` and run the rest of the test sequentially as it was before? I don't see the advantage of running all this in a `thenCompose()`.

Reply:
Yeah. Removed thenCompose and made it similar to before.